lms/djangoapps/courseware/block_render.py

```diff
@@ -364,60 +370,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
 # pylint: disable=too-many-statements
-def get_block_for_descriptor(user, request, block, field_data_cache, course_key,
-                             position=None, wrap_xblock_display=True, grade_bucket_type=None,
-                             static_asset_path='', disable_staff_debug_info=False,
-                             course=None, will_recheck_access=False):
+def get_block_for_descriptor(
+    user: User,
+    request: Request | None,
+    block: XBlock,
+    field_data_cache: FieldDataCache | None,
+    course_key: CourseKey,
+    position: int | None = None,
+    wrap_xblock_display: bool = True,
+    grade_bucket_type: str | None = None,  # TODO: This is unused.
+    static_asset_path: str = '',
+    disable_staff_debug_info: bool = False,
+    course: CourseBlock | None = None,
+    will_recheck_access: bool = False,
+    track_function: Callable[[str, dict], None] | None = None,
+    student_data: KvsFieldData | None = None,
+    request_token: str | None = None,
+    user_location: str | None = None,
+) -> XBlock | None:
     """
     Implements get_block, extracting out the request-specific functionality.
 
     disable_staff_debug_info : If this is True, exclude staff debug information in the rendering of the block.
 
     See get_block() docstring for further details.
     """
-    track_function = make_track_function(request)
-
-    user_location = getattr(request, 'session', {}).get('country_code')
-
-    student_kvs = DjangoKeyValueStore(field_data_cache)
-    if is_masquerading_as_specific_student(user, course_key):
-        student_kvs = MasqueradingKeyValueStore(student_kvs, request.session)
-    student_data = KvsFieldData(student_kvs)
-
-    return get_block_for_descriptor_internal(
-        user=user,
-        block=block,
-        student_data=student_data,
-        course_id=course_key,
-        track_function=track_function,
-        position=position,
-        wrap_xblock_display=wrap_xblock_display,
-        grade_bucket_type=grade_bucket_type,
-        static_asset_path=static_asset_path,
-        user_location=user_location,
-        request_token=xblock_request_token(request),
-        disable_staff_debug_info=disable_staff_debug_info,
-        course=course,
-        will_recheck_access=will_recheck_access,
+    if request:
+        track_function = track_function or make_track_function(request)
+        user_location = user_location or getattr(request, 'session', {}).get('country_code')
+        request_token = request_token or xblock_request_token(request)
+
+    if not student_data:
+        student_kvs = DjangoKeyValueStore(field_data_cache)
+        if is_masquerading_as_specific_student(user, course_key):
+            student_kvs = MasqueradingKeyValueStore(student_kvs, request.session)
+        student_data = KvsFieldData(student_kvs)
+
+    # The runtime is already shared between XBlocks. If there are no wrappers, it is the first initialization.
+    should_recreate_runtime = not block.runtime.wrappers
```

Comment:
We are not checking services because some of them are initialized by the `Runtime` class from the `XBlock` library.